keps/20200901-standard-crd.md

```diff
@@ -0,0 +1,277 @@
+      data:
+        config.yml: |
+          endpoints:
+          - https://{{ .data.user }}:{{ .data.password }}@api.exmaple.com
+
+      # Uses an existing template from configmap
+      # secret is fetched, merged and templated within the referenced configMap data
+      # It does not update the configmap, it creates a secret with: data["alertmanager.yml"] = ...result...
+      templateFrom:
+      - configMap:
+          name: alertmanager
+          items:
+          - key: alertmanager.yaml
+
+  # data contains key/value pairs which correspond to the keys in the resulting secret
+  data:
+
+    # EXAMPLE 1: simple mapping
+    # one key from a store may hold multiple values
+    # we need a way to map the values to the frontend
+    # it is the responsibility of the store implementation to know how to extract a value
+    tls.crt:
+      key: /corp.org/dev/certs/ingress
+      property: pubcert
+    tls.key:
+      key: /corp.org/dev/certs/ingress
+      property: privkey
+
+  # used to fetch all properties from a secret.
+  # if multiple dataFrom are specified, secrets are merged in the specified order
+  dataFrom:
+  - key: /user/all-creds
+
+status:
+  # represents the current phase of secret sync:
+  # * Pending | ES created, controller did not yet sync the ES or other dependencies are missing (e.g. secret store or configmap template)
+  # * Syncing | ES is being actively synced according to spec
+  # * Failing | Secret can not be synced, this might require user intervention
+  # * Failed  | ES can not be synced right now and will not able to
+  # * Completed | ES was synced successfully (one-time use only)
+  phase: Syncing
+  conditions:
+  - type: InSync
+    status: "True" # False if last sync was not successful
+    reason: "SecretSynced"
+    message: "Secret was synced"
+    lastTransitionTime: "2019-08-12T12:33:02Z"
+    lastSyncTime: "2020-09-23T16:27:53Z"
+
+```
+
+#### Behavior
+
+The ExternalSecret control loop **ensures** that the target resource exists and stays up to date with the upstream provider. Because most upstream APIs are limited in throughput the control loop must implement some sort of jitter and retry/backoff mechanic.
+
+### External Secret Store
+
+The store configuration in an `ExternalSecret` may contain a lot of redundancy, this can be factored out into its own CRD.
+These stores are defined in a particular namespace using `SecretStore` **or** globally with `GlobalSecretStore`.
+
+```yaml
+apiVerson: external-secrets.k8s.io/v1alpha1
+kind: SecretStore # or ClusterSecretStore
+metadata:
+  name: vault
+  namespace: example-ns
+spec:
+
+  # optional.
+  # used to select the correct KES controller (think: ingress.ingressClassName)
+  # The KES controller is instantiated with a specific controller name
+  # and filters ES based on this property
+  controller: "dev"
+
+  store:
+    # store implementation
+    vault:
```

Comment:
We've being playing with the idea of the secret store here at CS, and we think that would be simpler to avoid having different keys for each store type.

So, instead of having:

```
vault:
      server: "https://vault.example.com"
      path: path/on/vault/store
      auth:
     .
     .
     .
# 
## or
# 
# gsm:
#     projectID: some-project-id
#     auth:
#     .
#     .
#     .
```

We could just have:

```
  store:
    type: vault
    parameters:
      server: "https://vault.example.com"
      path: path/on/vault/store
    auth: 
    .
    .
    .
  #
  ## or
  #
  # store:
  #   type: gsm
  #   parameters:
  #     projectID: external-secrets-operator
  #   auth: 
  #   .
  #   .
  #   .
```

Then different implementations, that ask for different parameters, can just pass different things. And we just take care in the implementation to be open to receive anything passed under parameters. I think it is also easier to support new store types and avoid having to change the CRD to have more keys for stores. What do you think?

Reply:
I had that thought, too. If we can guarantee that 1) the CRD is validated properly and 2) we are not limited to k/v pairs in the `parameters` field we should do that. But i think `kubebuilder/controller-tools` doesn't support that yet. I value CRD validation a lot. just my two cents.

Reply:
I don't think we will be limited by specific k/v pairs here.  @iamcaleberic What do you think about CRD validation here?

Reply:
@moolen the controller runtime does have a `runtime.RawExtension` that we are using so as to not be limited to key:value pairs https://godoc.org/k8s.io/apimachinery/pkg/runtime#RawExtension.

Reply:
@iamcaleberic as discussed with @knelasevero: It's not suuper important to me, let's do it!

I've added it to the CRD spec.